pkg/northbound/proto/admin.proto

```diff
@@ -19,6 +19,152 @@ syntax = "proto3";
+message RemoveDeviceRequest {
+    string device_id = 1;
+}
+
+// RemoveDeviceResponse is sent in response to a RemoveDeviceRequest
+message RemoveDeviceResponse {
+
+}
+
+// StreamDevicesRequest requests a stream of device changes
+// By default, the request requests a stream of all device events that occur *after* the request is
+// received by the service. However, if `all` is `true`, the stream is prefixed by a device `ADD` event
+// for each existing device in the store.
+message StreamDevicesRequest {
+
+    // all indicates whether to initially stream existing devices in the store
+    bool all = 1;
+}
+
+// StreamDevicesResponse carries a single device event
+message StreamDevicesResponse {
+
+    // type is the type of the event
+    Type type = 1;
+
+    // device is the device on which the event occurred
+    Device device = 2;
+
+    // Device event type
+    enum Type {
+        // ADD occurs when a device is added to the topology
+        ADD = 0;
+
+        // UPDATE occurs when a device is updated
+        UPDATE = 1;
+
+        // REMOVE occurs when a device is removed from the topology
+        REMOVE = 2;
+    }
+}
+
+// Device contains information about a device
+message Device {
+    // id is a globally unique device identifier
+    string id = 1;
+
+    // address is the host:port of the device
+    string address = 2;
+
+    // target is the device target
+    string target = 3;
+
+    // version is the device version
+    string version = 4;
+
+    // timeout indicates the device request timeout
+    int64 timeout = 5;
+
+    // credentials contains the credentials for connecting to the device
+    Credentials credentials = 6;
+
+    // tls is the device TLS configuration
+    TlsConfig tls = 7;
+}
+
+// Credentials is the device credentials
+message Credentials {
+
+    // user is the user with which to connect to the device
+    string user = 1;
+
+    // password is the password for connecting to the device
+    string password = 2;
+}
+
+// Device TLS configuration
+message TlsConfig {
+
+    // caCert is the name of the device's CA certificate
+    string caCert = 3;
+
+    // cert is the name of the device's certificate
+    string cert = 4;
+
+    // key is the name of the device's TLS key
+    string key = 5;
+
+    // plain indicates whether to connect to the device over plaintext
+    bool plain = 6;
+
+    // insecure indicates whether to connect to the device with insecure communication
+    bool insecure = 7;
+}
+
 // TopoAdminService provides means for interactions with the topology subsystem.
 service TopoAdminService {
+
+    // AddDevice adds a device to the topology
+    rpc AddDevice(AddDeviceRequest) returns (AddDeviceResponse) {
+    }
+
+    // GetDevice gets a device by ID
+    rpc GetDevice(GetDeviceRequest) returns (GetDeviceResponse) {
+    }
+
+    // GetDevices gets a list of devices
+    rpc GetDevices(GetDevicesRequest) returns (GetDevicesResponse) {
```

Comment:
the reply here could be stream too (rather than a synchronous response)

Reply:
Agreed, the response should be a stream, as it makes reading a large number of devices much more efficient. Probably the use cases we are trying to cover do not include a large number of devices, but it seems a nice design decision that would reduce tech debt. Especially if we go down the path of using generic Read RPC (cc @tomikazi). Say one day we will need to get Links, or Ports, or Channels (of a port)... that can easily become a large number. I am strongly in favor of using a generic Read RPC (with type `DEVICE`, 'LINK', etc) as we automatically enforce the same good design for every read operation. instead of worrying whether we need a stream response or not.

Reply:
Yes, I completely agree that the get response should be a stream. I also agree that there will be additional entities down the line. I was envisioning the topology subsystem as a collection of services for tracking devices, their ports, links, networks.

I'm not opposed to collapsing this, but reductio ad absurdum will lead us to an API with mere GET/PUT. To me this is a trade-off in readability and maintanability. The repeated patterns in the methods in the various inventory services (device, link, networks, etc.) are a down-side, but on the other side where we have a common facade for accessing everything will be an intertwined and complex set of messages and schemes to differentiate one type of entity from another. In my view, while it will admittedly result in a slightly repeated pattern between devices, link, networks, etc., it will also give independence to the different services to properly cater to their respective subjects, in addition to making things more readable.

Reply:
It's not really clear to me what the benefit of streaming the list of devices is. Unless we anticipate the device list as being so large that it has the potential to impact memory/efficiency, streaming doesn't make sense for a request-reply message that's returning a discrete set of objects. This is _why_ the `repeated` field exists. Streaming the response increases the complexity of clients which then have to aggregate a response that really should have already been aggregated by the gRPC framework. Streams should be reserved for data that is a legitimate _stream_, and when they're not the specification is just describing a discrete set as a continuous one - describing the API and increasing the complexity of clients - without much benefit that I can see.

And I tend to agree with @tomikazi on the point that reducing the RPCs using `oneof` creates a more complicated message structure. What is the benefit beyond reducing the total number of RPCs, which seem to be basically free? I would suggest also that the API should be a collection of services as opposed to a single service for all entities.

Reply:
The benefit is scalability. Using streams now allow us to consider the case where we need to return and process a large number of devices (and links, ports, channels, etc) from the get-go. Moreover, I think the additional complexity at the client for the aggregation part is negligible, and shouldn't be a concern if this design favors scalability. What is more, grpc-java offers natively the possibility of wrapping a streaming RPC like this one in a sync stub, handling the aggregation of responses internally. I believe other grpc implementations for other languages provide the same capability. If that's not the case, apps can always implement their internal libraries to perform aggregation and re-use that.

Using generic Read/Write RPC with `oneof` field(s) is a way to follow the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle. I sketched here what I have in mind: https://gist.github.com/ccascone/32fc537f70c56ab927b67f5c69444d74

Does it make sense to apply it if all we need to manage is 3 different entities? Maybe not, what about 10? or 20? Perhaps knowing the type of entities that need to be managed would help us make an informed decision.

Reply:
These are all good points. I need to think about it a bit more. It could be my intuition is off. :)

Reply:
My two cents. 

1. I definitely like the streaming, independently of the approach we choose for the API methods/entities, so that should be adopted. 
2. Regarding the entities and method calls discussion. I do like https://gist.github.com/ccascone/32fc537f70c56ab927b67f5c69444d74 from @ccascone.
We can go for this solution if we think a first class object is an optical link vs a ran link for example. 
In this case the number of entities I think it will be in the order of 30-50 considering IP, Optical, RAN traffic. 
if instead we prefer to have a generic `link` first class object and then have `link_type` attribute the number of object will probably reduce to 10~, at least that's my impression. 
My preference would be to be as explicit as possible, thus creating first class objects for different types of devices, links etc. thus picking the solution Carmelo proposed.